packages/nimble/inst/tests/test-externalCalls.R

```diff
@@ -0,0 +1,98 @@
+## This is not an active test file.
+## It is really a demo with a couple of tests shown at the end.
+## It covers rough draft functionality for calling external compiled functions and calling arbitrary R functions from within compiled nimbleFunction (including BUGS) code.
+
+source(system.file(file.path('tests', 'test_utils.R'), package = 'nimble'))
+
+message('Ignore warnings from DSL checking. That system has not been updated for some new internal keywords used for external calls.')
+
+## Calling external C:
+
+## Say we want a C function that adds 1.5 to a vector of values:
+## We can only give non-scalars results by pointer argument.
+## Any NIMBLE non-scalar can become a double* of contiguously allocated memory.
+## Like in LAPACK or similar low-level libraries, you need a separate argument to say how long the allocated memory is.
+sink('add1p5.h')
+cat('
+extern "C" {
+  void my_internal_function(double *p, double*ans, int n);
+}
+')
+sink()
+
+sink('add1p5.cpp') 
+cat('
+#include <cstdio>
+#include "add1p5.h"
+
+void my_internal_function(double *p, double *ans, int n) {
+  printf("In my_internal_function\\n"); /* cat reduces the double slash to single slash */ 
+  for(int i = 0; i < n; i++) 
+    ans[i] = p[i] + 1.5;
+}
+')
+sink()
+
+## A major limitation is that we cannot have compiled code return anything except a scalar.  So returned non-scalars will have to be via arguments.  That meansto use this in BUGS code we'll have to wrap it in another nimbleFunction.  It will make sense below.
```

Comment:
Sounds like an opportunity for a convenience wrapper or an optional argument `returnArg` or somethting :grinning: